contracts/external/policies/TimeFramePolicy.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.23;
+
+import "../../DataTypes.sol";
+import { IActionPolicy, I1271Policy, IUserOpPolicy, IPolicy, VALIDATION_SUCCESS, VALIDATION_FAILED } from "../../interfaces/IPolicy.sol";
+import { PackedUserOperation, _packValidationData } from "@rhinestone/modulekit/src/external/ERC4337.sol";
+import { IERC165 } from "forge-std/interfaces/IERC165.sol";
+
+struct TimeFrameConfig {
+    uint48 validUntil;
+    uint48 validAfter;
+}
+
+
+/**
+ * @title TimeFramePolicy
+ * @notice This policy checks if current block timestamp is within a valid time frame.
+ * It is used to limit the time frame of an action / user operation.
+ * It packs the valid time frame into validation data to be returned to the EntryPoint.
+ */
+contract TimeFramePolicy is IPolicy, IUserOpPolicy, IActionPolicy, I1271Policy {
+
+    mapping(ConfigId id => mapping(address msgSender => mapping(address opSender => TimeFrameConfig))) public
+        timeFrameConfigs;
+
+    /**
+     * @notice Checks if the user operation is within the valid time frame.
+     * @param id The config ID.
+     * @param op The user operation.
+     */
+    function checkUserOpPolicy(ConfigId id, PackedUserOperation calldata op) external view override returns (uint256) {
+        return _checkTimeFrame(id, msg.sender, op.sender);
+    }
+
+    /**
+     * @notice Checks if the action is within the valid time frame.
+     * @param id The config ID.
+     * @param account The account.
+     */
+    function checkAction(
+        ConfigId id,
+        address account,
+        address,
+        uint256,
+        bytes calldata
+    )
+        external
+        view
+        returns (uint256)
+    {
+        return _checkTimeFrame(id, msg.sender, account);
+    }
+
+    /**
+     * @notice Checks if the 1271 signed action is within the valid time frame.
+     * @param id The config ID.
+     * @param smartAccount The account.
+     */
+    function check1271SignedAction(
+        ConfigId id,
+        address,
+        address smartAccount,
+        bytes32,
+        bytes calldata
+    )
+        external
+        view
+        returns (bool)
+    {
+        TimeFrameConfig memory config = timeFrameConfigs[id][msg.sender][smartAccount];
+        require(config.validUntil != 0 || config.validAfter != 0, PolicyNotInitialized(id, msg.sender, smartAccount));
+        if ((block.timestamp < config.validUntil || config.validUntil == 0) && block.timestamp >= config.validAfter) {
```

Comment:
should we also allow validAfter to be 0?

Reply:
if you mean line 73, it is implicitly allowed. As with validAfter == 0, `block.timestamp >= config.validAfter` is always true